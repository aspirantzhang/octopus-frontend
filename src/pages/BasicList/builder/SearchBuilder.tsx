import React from 'react';
import moment from 'moment';
import { Input, InputNumber, Form, DatePicker, TreeSelect, Col, Select, Radio } from 'antd';

const SearchBuilder = (data: BasicListApi.Field[] | undefined) => {
  return (Array.isArray(data) ? data : []).map((field) => {
    const formItemAttr = {
      label: field.title,
      name: field.name,
      key: field.name,
    };
    switch (field.type) {
      case 'input':
        return (
          <Col sm={6} key={field.name}>
            <Form.Item {...formItemAttr}>
              <Input />
            </Form.Item>
          </Col>
        );
      case 'textarea':
        return (
          <Col sm={6} key={field.name}>
            <Form.Item {...formItemAttr}>
              <Input />
            </Form.Item>
          </Col>
        );
      case 'number':
        return (
          <Col sm={6} key={field.name}>
            <Form.Item {...formItemAttr}>
              <InputNumber style={{ width: '100%' }} />
            </Form.Item>
          </Col>
        );
      case 'datetime':
        return (
          <Col sm={12} key={field.name}>
            <Form.Item {...formItemAttr}>
              <DatePicker.RangePicker
                showTime
                style={{ width: '100%' }}
                ranges={{
                  Today: [moment().startOf('day'), moment().endOf('day')],
                  'Last 7 Days': [moment().subtract(7, 'd'), moment()],
                  'Last 30 Days': [moment().subtract(30, 'days'), moment()],
                  'Last Month': [
                    moment().subtract(1, 'months').startOf('month'),
                    moment().subtract(1, 'months').endOf('month'),
                  ],
                }}
              />
            </Form.Item>
          </Col>
        );
      case 'switch':
        return (
          <Col sm={6} key={field.name}>
            <Form.Item {...formItemAttr} valuePropName="checked">
              <Select>
                {(Array.isArray(field.data) ? field.data : []).map((option: any) => {
                  return (
                    <Select.Option value={option.value} key={option.value}>
                      {option.title}
                    </Select.Option>
                  );
                })}
              </Select>
            </Form.Item>
          </Col>
        );
      case 'radio':
        return (
          <Col sm={6} key={field.name}>
            <Form.Item {...formItemAttr}>
              <Radio.Group buttonStyle="solid">
                {(field.data || []).map((item: any) => {
                  return (
                    <Radio.Button key={item.value} value={item.value}>
                      {item.title}
                    </Radio.Button>
                  );
                })}
              </Radio.Group>
            </Form.Item>
          </Col>
        );
      case 'category':
      case 'tree':
        return (
          <Col sm={6} key={field.name}>
            <Form.Item {...formItemAttr}>
              <TreeSelect treeData={field.data} treeCheckable />
            </Form.Item>
          </Col>
        );
      case 'parent':
        return (
          <Col sm={6} key={field.name}>
            <Form.Item {...formItemAttr}>
              <TreeSelect treeData={field.data} treeDefaultExpandAll allowClear />
            </Form.Item>
          </Col>
        );
      case 'select':
        return (
          <Col sm={6} key={field.name}>
            <Form.Item {...formItemAttr} valuePropName="checked">
              <Select>
                {(Array.isArray(field.data) ? field.data : []).map((option: any) => {
                  return (
                    <Select.Option value={option.value} key={option.value}>
                      {option.title}
                    </Select.Option>
                  );
                })}
              </Select>
            </Form.Item>
          </Col>
        );

      default:
        return null;
    }
  });
};

export default SearchBuilder;
