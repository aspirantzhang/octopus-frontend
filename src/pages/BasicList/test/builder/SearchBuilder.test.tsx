/**
 * @jest-environment jsdom
 */
import './antd-test-setup';
import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render, screen } from '@testing-library/react';
import { Form } from 'antd';
import SearchBuilder from '../../builder/SearchBuilder';

const validSearch = [
  {
    name: 'single_line_text',
    title: 'Single-line Text',
    type: 'input',
    disabled: true,
  },
  {
    name: 'password',
    title: 'Password',
    type: 'password',
    disabled: true,
  },
  {
    name: 'multi_line_text',
    title: 'Multi-line Text',
    type: 'textarea',
    disabled: true,
  },
  {
    name: 'number',
    title: 'Number',
    type: 'number',
    disabled: true,
  },
  {
    name: 'datetime',
    title: 'Datetime',
    type: 'datetime',
    disabled: true,
  },
  {
    name: 'tree',
    title: 'Tree',
    type: 'tree',
    data: [],
    disabled: true,
  },
  {
    name: 'select',
    title: 'Select',
    type: 'select',
    data: [],
    disabled: true,
  },
  {
    name: 'switch',
    title: 'Switch',
    type: 'switch',
    data: [{ title: 'foo', value: 'bar' }],
    disabled: true,
  },
  {
    name: 'invalid',
    title: 'Invalid',
    type: 'invalid',
  },
];

describe('SearchBuilder', () => {
  test('snapshot test', () => {
    const { container } = render(<Form>{SearchBuilder(validSearch as any)}</Form>);
    expect(container).toMatchSnapshot();
  });

  test('valid params', () => {
    render(<Form>{SearchBuilder(validSearch as any)}</Form>);
    expect(screen.getByLabelText('Single-line Text')).toBeInTheDocument();
    expect(screen.getByLabelText('Multi-line Text')).toBeInTheDocument();
    expect(screen.getByLabelText('Number')).toBeInTheDocument();
    expect(screen.getByLabelText('Datetime')).toBeInTheDocument();
    expect(screen.getByLabelText('Tree')).toBeInTheDocument();
    expect(screen.getByLabelText('Select')).toBeInTheDocument();
    expect(screen.getByLabelText('Switch')).toBeInTheDocument();
    expect(screen.queryByLabelText('Invalid')).not.toBeInTheDocument();
    expect(screen.queryByLabelText('Password')).not.toBeInTheDocument();
  });

  test('invalid params', () => {
    expect(SearchBuilder(null as any)).toEqual([]);
    expect(SearchBuilder([] as any)).toEqual([]);
    expect(SearchBuilder({} as any)).toEqual([]);
    expect(SearchBuilder(undefined as any)).toEqual([]);
    expect(SearchBuilder(NaN as any)).toEqual([]);
    expect(SearchBuilder(true as any)).toEqual([]);
    expect(SearchBuilder(false as any)).toEqual([]);
    expect(SearchBuilder('' as any)).toEqual([]);
  });
});
