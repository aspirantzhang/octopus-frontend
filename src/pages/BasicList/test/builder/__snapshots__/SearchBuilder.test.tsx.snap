// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBuilder snapshot test 1`] = `
<div>
  <form
    class="ant-form ant-form-horizontal"
  >
    <div
      class="ant-col ant-col-sm-6"
    >
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-label"
        >
          <label
            class=""
            for="single_line_text"
            title="Single-line Text"
          >
            Single-line Text
          </label>
        </div>
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <input
                class="ant-input"
                id="single_line_text"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ant-col ant-col-sm-6"
    >
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-label"
        >
          <label
            class=""
            for="multi_line_text"
            title="Multi-line Text"
          >
            Multi-line Text
          </label>
        </div>
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <input
                class="ant-input"
                id="multi_line_text"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ant-col ant-col-sm-6"
    >
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-label"
        >
          <label
            class=""
            for="number"
            title="Number"
          >
            Number
          </label>
        </div>
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <div
                class="ant-input-number"
                style="width: 100%;"
              >
                <div
                  class="ant-input-number-handler-wrap"
                >
                  <span
                    aria-disabled="false"
                    aria-label="Increase Value"
                    class="ant-input-number-handler ant-input-number-handler-up"
                    role="button"
                    unselectable="on"
                  >
                    <span
                      aria-label="up"
                      class="anticon anticon-up ant-input-number-handler-up-inner"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="up"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M890.5 755.3L537.9 269.2c-12.8-17.6-39-17.6-51.7 0L133.5 755.3A8 8 0 00140 768h75c5.1 0 9.9-2.5 12.9-6.6L512 369.8l284.1 391.6c3 4.1 7.8 6.6 12.9 6.6h75c6.5 0 10.3-7.4 6.5-12.7z"
                        />
                      </svg>
                    </span>
                  </span>
                  <span
                    aria-disabled="false"
                    aria-label="Decrease Value"
                    class="ant-input-number-handler ant-input-number-handler-down"
                    role="button"
                    unselectable="on"
                  >
                    <span
                      aria-label="down"
                      class="anticon anticon-down ant-input-number-handler-down-inner"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="down"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                        />
                      </svg>
                    </span>
                  </span>
                </div>
                <div
                  class="ant-input-number-input-wrap"
                >
                  <input
                    autocomplete="off"
                    class="ant-input-number-input"
                    id="number"
                    role="spinbutton"
                    step="1"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ant-col ant-col-sm-12"
    >
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-label"
        >
          <label
            class=""
            for="datetime"
            title="Datetime"
          >
            Datetime
          </label>
        </div>
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <div
                class="ant-picker ant-picker-range"
                style="width: 100%;"
              >
                <div
                  class="ant-picker-input ant-picker-input-active"
                >
                  <input
                    autocomplete="off"
                    id="datetime"
                    placeholder="Start date"
                    readonly=""
                    size="21"
                    value=""
                  />
                </div>
                <div
                  class="ant-picker-range-separator"
                >
                  <span
                    aria-label="to"
                    class="ant-picker-separator"
                  >
                    <span
                      aria-label="swap-right"
                      class="anticon anticon-swap-right"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="swap-right"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="0 0 1024 1024"
                        width="1em"
                      >
                        <path
                          d="M873.1 596.2l-164-208A32 32 0 00684 376h-64.8c-6.7 0-10.4 7.7-6.3 13l144.3 183H152c-4.4 0-8 3.6-8 8v60c0 4.4 3.6 8 8 8h695.9c26.8 0 41.7-30.8 25.2-51.8z"
                        />
                      </svg>
                    </span>
                  </span>
                </div>
                <div
                  class="ant-picker-input"
                >
                  <input
                    autocomplete="off"
                    placeholder="End date"
                    readonly=""
                    size="21"
                    value=""
                  />
                </div>
                <div
                  class="ant-picker-active-bar"
                  style="left: 0px; width: 0px; position: absolute;"
                />
                <span
                  class="ant-picker-suffix"
                >
                  <span
                    aria-label="calendar"
                    class="anticon anticon-calendar"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="calendar"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M880 184H712v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H384v-64c0-4.4-3.6-8-8-8h-56c-4.4 0-8 3.6-8 8v64H144c-17.7 0-32 14.3-32 32v664c0 17.7 14.3 32 32 32h736c17.7 0 32-14.3 32-32V216c0-17.7-14.3-32-32-32zm-40 656H184V460h656v380zM184 392V256h128v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h256v48c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8v-48h128v136H184z"
                      />
                    </svg>
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ant-col ant-col-sm-6"
    >
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-label"
        >
          <label
            class=""
            for="switch"
            title="Switch"
          >
            Switch
          </label>
        </div>
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <div
                class="ant-select ant-select-single ant-select-show-arrow"
              >
                <div
                  class="ant-select-selector"
                >
                  <span
                    class="ant-select-selection-search"
                  >
                    <input
                      aria-activedescendant="switch_list_0"
                      aria-autocomplete="list"
                      aria-controls="switch_list"
                      aria-haspopup="listbox"
                      aria-owns="switch_list"
                      autocomplete="off"
                      class="ant-select-selection-search-input"
                      id="switch"
                      readonly=""
                      role="combobox"
                      style="opacity: 0;"
                      type="search"
                      unselectable="on"
                      value=""
                    />
                  </span>
                  <span
                    class="ant-select-selection-placeholder"
                  />
                </div>
                <span
                  aria-hidden="true"
                  class="ant-select-arrow"
                  style="user-select: none;"
                  unselectable="on"
                >
                  <span
                    aria-label="down"
                    class="anticon anticon-down ant-select-suffix"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="down"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                      />
                    </svg>
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ant-col ant-col-sm-6"
    >
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-label"
        >
          <label
            class=""
            for="tree"
            title="Tree"
          >
            Tree
          </label>
        </div>
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <div
                class="ant-select ant-tree-select ant-select-multiple ant-select-show-search"
              >
                <div
                  class="ant-select-selector"
                >
                  <div
                    class="ant-select-selection-overflow"
                  >
                    <div
                      class="ant-select-selection-overflow-item ant-select-selection-overflow-item-suffix"
                      style="opacity: 1;"
                    >
                      <div
                        class="ant-select-selection-search"
                        style="width: 0px;"
                      >
                        <input
                          aria-activedescendant="tree_list_0"
                          aria-autocomplete="list"
                          aria-controls="tree_list"
                          aria-haspopup="listbox"
                          aria-owns="tree_list"
                          autocomplete="off"
                          class="ant-select-selection-search-input"
                          id="tree"
                          readonly=""
                          role="combobox"
                          style="opacity: 0;"
                          type="search"
                          unselectable="on"
                          value=""
                        />
                        <span
                          aria-hidden="true"
                          class="ant-select-selection-search-mirror"
                        >
                          
                           
                        </span>
                      </div>
                    </div>
                  </div>
                  <span
                    class="ant-select-selection-placeholder"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ant-col ant-col-sm-6"
    >
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-label"
        >
          <label
            class=""
            for="parent"
            title="Parent"
          >
            Parent
          </label>
        </div>
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <div
                class="ant-select ant-tree-select ant-select-single ant-select-allow-clear ant-select-show-arrow"
              >
                <div
                  class="ant-select-selector"
                >
                  <span
                    class="ant-select-selection-search"
                  >
                    <input
                      aria-activedescendant="parent_list_0"
                      aria-autocomplete="list"
                      aria-controls="parent_list"
                      aria-haspopup="listbox"
                      aria-owns="parent_list"
                      autocomplete="off"
                      class="ant-select-selection-search-input"
                      id="parent"
                      readonly=""
                      role="combobox"
                      style="opacity: 0;"
                      type="search"
                      unselectable="on"
                      value=""
                    />
                  </span>
                  <span
                    class="ant-select-selection-placeholder"
                  />
                </div>
                <span
                  aria-hidden="true"
                  class="ant-select-arrow"
                  style="user-select: none;"
                  unselectable="on"
                >
                  <span
                    aria-label="down"
                    class="anticon anticon-down ant-select-suffix"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="down"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                      />
                    </svg>
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ant-col ant-col-sm-6"
    >
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-label"
        >
          <label
            class=""
            for="select"
            title="Select"
          >
            Select
          </label>
        </div>
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <div
                class="ant-select ant-select-single ant-select-show-arrow"
              >
                <div
                  class="ant-select-selector"
                >
                  <span
                    class="ant-select-selection-search"
                  >
                    <input
                      aria-activedescendant="select_list_0"
                      aria-autocomplete="list"
                      aria-controls="select_list"
                      aria-haspopup="listbox"
                      aria-owns="select_list"
                      autocomplete="off"
                      class="ant-select-selection-search-input"
                      id="select"
                      readonly=""
                      role="combobox"
                      style="opacity: 0;"
                      type="search"
                      unselectable="on"
                      value=""
                    />
                  </span>
                  <span
                    class="ant-select-selection-placeholder"
                  />
                </div>
                <span
                  aria-hidden="true"
                  class="ant-select-arrow"
                  style="user-select: none;"
                  unselectable="on"
                >
                  <span
                    aria-label="down"
                    class="anticon anticon-down ant-select-suffix"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="down"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                      />
                    </svg>
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</div>
`;
